# Human Pseudoautosomal Boundary Analysis
### Daniel Cotter
#### 2018

## Data

### Variant Data
Variant data is from The 1000 Genomes Project phase3 VCF files for chrX, chrY, & chr8

- **chrX:** [ALL.chrX.phase3_shapeit2_mvncall_integrated_v1b.20130502.genotypes.vcf.gz](http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chrX.phase3_shapeit2_mvncall_integrated_v1b.20130502.genotypes.vcf.gz "chrX")

- **chrY:** [ALL.chrY.phase3_integrated_v2a.20130502.genotypes.vcf.gz](http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chrY.phase3_integrated_v2a.20130502.genotypes.vcf.gz "chrY")

- **chr8:** [ALL.chr8.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz](http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr8.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz "chr8")

### Genome masks
The strict mask as provided by The 1000 genomes Project is used for identifying monomorphic sites

- **Strict Mask chrX:** [20141020.chrX.strict_mask.fasta.gz](http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/accessible_genome_masks/StrictMask/20141020.chrX.strict_mask.fasta.gz "chrX")

- **Strict Mask chrY:** [20141020.chrY.strict_mask.fasta.gz](http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/accessible_genome_masks/StrictMask/20141020.chrY.strict_mask.fasta.gz "chrY")

- **Strict Mask chr8:** [20140520.chr8.strict_mask.fasta.gz](http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/accessible_genome_masks/StrictMask/20140520.chr8.strict_mask.fasta.gz "chr8")

### Population Lists
Population and subpopulation lists are calculated using [integrated_call_samples_v3.20130502.ALL.panel](http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/integrated_call_samples_v3.20130502.ALL.panel "Population panel")

#### Population Tables

#### Super Populations

POP | Females | Males
---|:---:|:---:
AFR|342|319
AMR|177|170
EAS|260|244
EUR|263|240
SAS|229|260

#### Populations

POP | Females | Males
---|:---:|:---:
ACB|49|47
ASW|35|26
BEB|44|42
CDX|49|44
CEU|50|49
CHB|57|46
CHS|53|52
CLM|51|43
ESN|46|53
FIN|61|38
GBR|45|46
GIH|47|56
GWD|58|55
IBS|53|54
ITU|43|59
JPT|48|56
KHV|53|46
LWK|55|44
MSL|43|42
MXL|32|32
PEL|44|41
PJL|48|48
PUR|50|54
STU|47|55
TSI|54|53
YRI|56|52

Population codes can be found [here](http://www.internationalgenome.org/faq/which-populations-are-part-your-study/).

## Analysis

### Step 1: Parse populations into lists

- Item 1
- Item 2

### Step 2: Compute diversity by population

- Item 1
- Item 2
